File: score/test_scenarios_cpp/examples/basic.cpp
// *******************************************************************************
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0
//
// SPDX-License-Identifier: Apache-2.0
// *******************************************************************************

#include <cli.hpp>
#include <scenario.hpp>
#include <test_context.hpp>
#include <tracing.hpp>

#include <nlohmann/json.hpp>

#include <cstddef>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kTargetName{"examples::basic::list"};

std::vector<std::string> parse_items(const std::string& input) {
    nlohmann::json parsed;
    try {
        parsed = nlohmann::json::parse(input);
    } catch (const nlohmann::json::exception& e) {
        throw std::runtime_error{"invalid input: " + std::string{e.what()}};
    }
    if (!parsed.contains("items")) {
        throw std::runtime_error{"invalid input: missing 'items' field"};
    }
    return parsed.at("items").get<std::vector<std::string>>();
}

// Logs each item with its index via structured tracing.
class EnumerateScenario final : public Scenario {
   public:
    std::string name() const override { return "enumerate"; }

    void run(const std::string& input) const override {
        auto items{parse_items(input)};
        for (std::size_t index = 0; index < items.size(); ++index) {
            TRACING_INFO(kTargetName, std::pair{std::string{"index"}, index},
                         std::pair{std::string{"item"}, items[index]});
        }
    }
};

// Fails if the input has no items.
class RequireNonEmptyScenario final : public Scenario {
   public:
    std::string name() const override { return "require_non_empty"; }

    void run(const std::string& input) const override {
        auto items{parse_items(input)};
        if (items.empty()) {
            throw std::runtime_error{"items must not be empty"};
        }
        TRACING_INFO(kTargetName, std::pair{std::string{"count"}, items.size()});
    }
};

}  // namespace

int main(int argc, char* argv[]) {
    std::vector<std::string> raw_arguments{argv, argv + argc};

    ScenarioGroup::Ptr list_group{new ScenarioGroupImpl{
        "list",
        std::vector<Scenario::Ptr>{std::make_shared<EnumerateScenario>(),
                                    std::make_shared<RequireNonEmptyScenario>()},
        std::vector<ScenarioGroup::Ptr>{}}};
    ScenarioGroup::Ptr root_group{new ScenarioGroupImpl{
        "root", std::vector<Scenario::Ptr>{}, std::vector<ScenarioGroup::Ptr>{list_group}}};
    TestContext test_context{root_group};

    try {
        run_cli_app(raw_arguments, test_context);
        return 0;
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }
}

File: score/test_scenarios_rust/examples/basic.rs
// *******************************************************************************
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache License Version 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0
//
// SPDX-License-Identifier: Apache-2.0
// *******************************************************************************

use std::process::ExitCode;

use serde::Deserialize;
use tracing::info;

use test_scenarios_rust::cli::{create_tracing_subscriber, run_cli_app};
use test_scenarios_rust::scenario::{Scenario, ScenarioGroupImpl};
use test_scenarios_rust::test_context::TestContext;

#[derive(Deserialize)]
struct ItemsInput {
    items: Vec<String>,
}

impl ItemsInput {
    fn parse(input: &str) -> Result<Self, String> {
        serde_json::from_str(input).map_err(|e| format!("invalid input: {e}"))
    }
}

/// Logs each item with its index via structured tracing.
struct EnumerateScenario;

impl Scenario for EnumerateScenario {
    fn name(&self) -> &str {
        "enumerate"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let parsed = ItemsInput::parse(input)?;
        for (index, item) in parsed.items.iter().enumerate() {
            info!(index, item = item.as_str());
        }
        Ok(())
    }
}

/// Fails if the input has no items.
struct RequireNonEmptyScenario;

impl Scenario for RequireNonEmptyScenario {
    fn name(&self) -> &str {
        "require_non_empty"
    }

    fn run(&self, input: &str) -> Result<(), String> {
        let parsed = ItemsInput::parse(input)?;
        if parsed.items.is_empty() {
            return Err("items must not be empty".to_string());
        }
        info!(count = parsed.items.len());
        Ok(())
    }
}

fn main() -> ExitCode {
    tracing::subscriber::set_global_default(create_tracing_subscriber())
        .expect("Setting default subscriber failed!");

    let raw_arguments: Vec<String> = std::env::args().collect();

    let list_group = ScenarioGroupImpl::new(
        "list",
        vec![Box::new(EnumerateScenario), Box::new(RequireNonEmptyScenario)],
        Vec::new(),
    );
    let root_group = ScenarioGroupImpl::new("root", Vec::new(), vec![Box::new(list_group)]);
    let test_context = TestContext::new(Box::new(root_group));

    match run_cli_app(&raw_arguments, &test_context) {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("{e}");
            ExitCode::FAILURE
        }
    }
}
